fix(rust): support for `rust-toolchain.toml`

pineapplehunter opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
<!--- Describe your changes in detail -->
This PR adds support for `rust-toolchain.toml` in the rust module.
A related Issue is #417, and this just adds some lines to PR #426.

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
I haven't opened a issue for this, but when I enter a directory with `rust-toolchain.toml` rustup started to install the toolchain. This is the same issue as #417 but with `.toml` file.

#### Screenshots (if appropriate):

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

davidkna commented:
Thanks for the contribution @pineapplehunter, to get the CI to pass can you either update your commit message to include the prefix `fix(rust):` in the title or just push an empty commit to this branch e.g

```sh
git commit --allow-empty -m 'make Semantic PR bot happy'
git push 
```

pineapplehunter commented:
@davidkna I've changed the commit message.

davidkna commented:
This looks good to me, but could you maybe add a test for this new case?

pineapplehunter commented:
I added some tests!
When making a test a realized that when using the `.toml` extension, you can't use the channel in one line format. So I changed the `read_channel` function so it won't parse in that case.

pineapplehunter commented:
I had some time, so I added some more tests, and comments.
Found out that the code was checking the current directory twice, so I removed that part.
If the tests pass, I think this PR is done.

davidkna commented on src/modules/rust.rs:
```diff
@@ -137,12 +140,12 @@ fn find_rust_toolchain_file(context: &Context) -> Option<String> {
-        match contents.lines().count() {
-            0 => None,
-            1 => Some(contents),
+        match (contents.lines().count(), only_toml) {
+            (0, _) => None,
+            (1, false) => Some(contents),
```
I think something like this would be more clear:
```suggestion
        match contents.lines().count() {
            0 => None,
            1 if !only_toml => Some(contents),
```

pineapplehunter replied:
That looks good!

pineapplehunter commented:
It seems like clippy is failing on a unrelated to the PR

davidkna commented:
@pineapplehunter Please rebase on/merge master to make clippy pass again.

pineapplehunter commented:
I rebased the commits on master.
I think this should work.

pineapplehunter commented:
The error in the test seems unrelated to this PR, but I don't know.
I don't have a mac to check about this.
Is there something I can do to fix this?

davidkna commented:
I'm rerunning the tests now. I think this failure was just #886.

davidkna commented on src/modules/rust.rs:
```diff
@@ -154,18 +157,12 @@ fn find_rust_toolchain_file(context: &Context) -> Option<String> {
         .map(|c| c.trim().to_owned())
     }
 
-    if let Ok(true) = context
-        .dir_contents()
-        .map(|dir| dir.has_file("rust-toolchain"))
-    {
-        if let Some(toolchain) = read_channel(Path::new("rust-toolchain")) {
-            return Some(toolchain);
-        }
-    }
-
```
I just looked at the code again and saw that you removed this. Can you add this back again? Unlike the code below this will have the current directory cached in `context`.

pineapplehunter replied:
Got it. I'll fix that part.
But I've got some questions.
1. Should I copy this code for `rust-toolchain.toml` too? I'm gessing if this is for caching so this only needs to be run only once for `rust-toolnain`.
2. Should the loop below looking for parent directories start from a parent? If I add the code back, I think it would try to load the same file in the current directory twice.

davidkna replied:
Please use this for the toml file too. You can also simplify it to something likes this:

```rust
if context
        .dir_contents()
        .map_or(false, |dir| dir.has_file("rust-toolchain"))
```

I think it's fine to have it start from the parent (though that might end up causing issues with case-insensitive file systems).

pineapplehunter replied:
Thanks for the quick response!
I'll use that code.
I found out that starting from the parent in the loop causes a error in the test. It seems that
 
```rust
read_channel(Path::new("rust-toolchain"), false) // returns None, path = "rust-toolchain"
```

and

```rust
let mut dir = &*context.current_dir;
read_channel(&dir.join("rust-toolchain"), false) // returns Some, path = "/tmp/**something**/rust-toolchain"
```

give different outputs. So I'll keep the loop as is.

pineapplehunter commented:
Updated the code.
What do you think?

davidkna commented:
Thanks for the contribution @pineapplehunter!
